### DISCUSSION TITLE
Original exception never logged with Retryable Topics

### DISCUSSION THREAD

**Discussion by tomazfernandes** [spring-projects/spring-kafka#2211] [General]
This [SO](https://stackoverflow.com/questions/71719065/backoffexceptions-are-logged-at-error-level-when-using-retrytopicconfiguration) user pointed out that, when the record goes through the Retry Topic flow, the original exception never gets logged.

We had already discussed this a bit earlier this year. 

@garyrussell [proposed](https://github.com/spring-projects/spring-kafka/issues/2063#issuecomment-1010404269_):

> But I see your point; since we always recover these failures, they are never logged. Maybe the DLPR needs an option to log the exception after successfully recovering a record (but that doesn't need to be part of this fix).

I suggested as an [alternative](https://github.com/spring-projects/spring-kafka/issues/2063#issuecomment-1010482981_):

> We could log the error message just before we either send the record to the DLT or give up processing - that way nothing gets logged if any of the retrials work, and gets logged only once if none.

The main difference seems to be whether we want to log the exception at every retry attempt or only before it goes to the DLT.

We might look into adding both and making these configurable - users might be able to set a flag in `DLPR`  to log after every attempt, and `DLPRF` to log before going to the DLT.

We could also have this as an enum in `DLPRF`, with something like `OriginalExceptionLoggingStrategy.NEVER`, `EVERY_ATTEMPT`, `AFTER_RETRIES_EXHAUSTED` - `DLPRF` would then set the appropriate flag in `DLPR`.

WDYT?

**Comment by tomazfernandes** [comment]
Just to illustrate, if we decide to have the option of logging only after retry exhaustion, this is where I think we might do it:

https://github.com/spring-projects/spring-kafka/blob/f53d41b3c7c3572c1124d280641112d458bf6cf7/spring-kafka/src/main/java/org/springframework/kafka/retrytopic/DeadLetterPublishingRecovererFactory.java#L176-L191

If `nextDestination` is dlt or noOps, we'd log the exception.

**Reply by garyrussell** [reply]
Yes; that makes sense.

**Reply by tomazfernandes** [reply]
Ok, I think the easiest fix is to just to add this logic here - I don't think it really makes sense to never log the exception.

We can then consider adding an option of logging every attempt if necessary. I think it would be useful, but since we're still working on improving configuration for the feature as a while, perhaps it's one less thing to worry about configuring for now.

**Reply by tzie0062** [reply]
I'd also prefer the exception to only be logged once. However, I wonder about the right place to log, because the `dltHandler` can extract (and log) this information as well. So does this belong in the framework or in the app?

**Reply by tomazfernandes** [reply]
That's a fair point. I think that OOTB the exception should be logged somewhere by the framework - worst thing is an error happening silently.

One option would be extracting exception info and logging it in the `default DLT handler` - this way it would be logged by default by the framework, but if users add a `dltHandler` of their own, it's now their responsibility to log.

https://github.com/spring-projects/spring-kafka/blob/f53d41b3c7c3572c1124d280641112d458bf6cf7/spring-kafka/src/main/java/org/springframework/kafka/retrytopic/RetryTopicConfigurer.java#L428-L441

Even though, I wonder if some users might find confusing if exception logging is gone after adding a `dltHandler`, and also there's the option of not configuring a DLT, so not sure how users would handle that.

**Reply by tzie0062** [reply]
I think I like your idea of using a `OriginalExceptionLoggingStrategy` in the `DLPR` which could default to `AFTER_RETRIES_EXHAUSTED`, leaving users the option to handle logging within application code by setting it to `NEVER`.

Would it make sense to expose some metrics (e.g. failures) via micrometer as well?

**Reply by tomazfernandes** [reply]
Sure, I think the framework should be as flexible as possible to adapt to users' needs. I think we can have all logging logic handled in the `DLPRF` - no need to change anything in `DLPR` for this.

TBH, I'm not that familiar with micrometer, maybe that's a good chance to learn more about it. But it probably should be a separate issue so that one thing doesn't end up delaying the other.

I'd like to hear @garyrussell's thoughts on both the Enum and the `micrometer` if possible, then I can change the PR I opened with this.

Thanks for the input!

**Reply by garyrussell** [reply]
The enum solution sounds good to me.

Exceptions are already counted by all containers, with the tags including the listener id (bean name).

https://docs.spring.io/spring-kafka/docs/current/reference/html/#monitoring-listener-performance

**Reply by tzie0062** [reply]
@garyrussell just out of curiosity: if I have two retry topics and all attempts to process an event fail, what will the metrics show? One or three exceptions?

**Reply by garyrussell** [reply]
The container is not aware of the retry mechanisms; all 3 exceptions will be counted, but they will be tagged with different container ids.

**Comment by tomazfernandes** [comment]
https://github.com/spring-projects/spring-kafka/issues/2212
